base/abstractarray.jl

```diff
@@ -2042,3 +2042,37 @@ function hash(a::AbstractArray, h::UInt)
     !isequal(x2, x1) && (h = hash(x2, h))
     return h
 end
+
+## Reversed-order iteration for arrays and other collections.  Collections
+## should implement start/next/done etcetera
+"""
+    Reversed(itr)
+
+Given an iterator `itr`, then `Reversed(itr)` is an iterator over the
+same collection but in the reverse order.
+
+Not all iterator types `T` support reverse-order iteration.  If `T`
+doesn't, then iterating over `Reversed(itr::T)` will throw a [`MethodError`](@ref)
+because of the missing [`start`](@ref), [`next`](@ref), and [`done`](@ref)
+methods for `Reversed{T}`.
+"""
+struct Reversed{T}
+    iter::T
+end
+eltype(r::Reversed) = eltype(r.iter)
+length(r::Reversed) = length(r.iter)
+size(r::Reversed) = size(r.iter)
+iteratorsize(r::Reversed) = iteratorsize(r.iter)
+iteratoreltype(r::Reversed) = iteratoreltype(r.iter)
+last(r::Reversed) = first(r.iter) # the first shall be last
+first(r::Reversed) = last(r.iter) # and the last shall be first
+
+# reverse-order array iterators: assumes more-specialized Reversed for eachindex
+start(A::Reversed{<:AbstractArray}) = (@_inline_meta; iter = Reversed(eachindex(A.iter)); (iter, start(iter)))
+next(A::Reversed{<:AbstractArray}, i) = (@_propagate_inbounds_meta; (idx, s) = next(i[1], i[2]); (A.iter[idx], (i[1], s)))
+done(A::Reversed{<:AbstractArray}, i) = (@_propagate_inbounds_meta; done(i[1], i[2]))
+
+Reversed(R::AbstractRange) = reverse(R) # copying ranges is cheap
+Reversed(G::Generator) = Generator(G.f, Reversed(G.iter))
+Reversed(r::Reversed) = r.iter
+reverse(r::Reversed) = collect(r.iter)
```

Comment:
Using `collect` is kind of weird for strings, isn't it? Or am I missing something?